File: packages/core/src/shell/builtins/jobs.ts
/**
 * jobs - display status of jobs
 */

import { BuiltinCommand, ShellEnvironment, ShellOptions } from '../types.js';

export class JobsCommand implements BuiltinCommand {
  name = 'jobs';
  description = 'Display status of jobs';

  async execute(args: string[], env: ShellEnvironment, options: ShellOptions): Promise<number> {
    if (!options.enableJobControl) {
      console.error('jobs: job control not enabled');
      return 1;
    }

    const jobs = env.getAllJobs();
    
    if (jobs.length === 0) {
      // No jobs - silent return
      return 0;
    }

    // Parse options
    let showPids = false;
    let longFormat = false;
    
    for (const arg of args) {
      if (arg === '-l') {
        longFormat = true;
      } else if (arg === '-p') {
        showPids = true;
      } else if (arg.startsWith('-')) {
        console.error(`jobs: ${arg}: invalid option`);
        return 1;
      }
    }

    // Sort jobs by ID
    const sortedJobs = jobs.sort((a, b) => a.id - b.id);

    // Find the current and previous jobs
    const runningJobs = sortedJobs.filter(j => j.status === 'running');
    const currentJob = runningJobs[runningJobs.length - 1];
    const previousJob = runningJobs[runningJobs.length - 2];

    // Display each job
    for (const job of sortedJobs) {
      if (showPids) {
        // -p option: just show PIDs
        console.log(job.pid);
      } else {
        // Format: [job_id]+ status command
        let marker = ' ';
        if (job === currentJob) marker = '+';
        else if (job === previousJob) marker = '-';

        const status = this.formatStatus(job.status, job.exitCode);
        
        if (longFormat) {
          // -l option: include PID
          console.log(`[${job.id}]${marker} ${job.pid} ${status}\t${job.command}`);
        } else {
          console.log(`[${job.id}]${marker}  ${status}\t${job.command}`);
        }
      }
    }

    return 0;
  }

  private formatStatus(status: string, exitCode?: number): string {
    switch (status) {
      case 'running':
        return 'Running';
      case 'stopped':
        return 'Stopped';
      case 'done':
        return exitCode === 0 ? 'Done' : `Exit ${exitCode}`;
      default:
        return status;
    }
  }

  help(): string {
    return `jobs [-lp]
Display status of jobs.

Options:
  -l    List process IDs in addition to normal information
  -p    List process IDs only

The status of each job is displayed, along with its job number and command.
Jobs marked with + are the current job; jobs marked with - are the previous job.`;
  }
}

File: shell/src/shell/builtins/jobs.rs
//! jobs - list active jobs builtin command

use crate::shell::types::*;

pub struct JobsCommand;

#[async_trait::async_trait]
impl BuiltinCommand for JobsCommand {
    fn name(&self) -> &str {
        "jobs"
    }

    fn description(&self) -> &str {
        "Display status of jobs in current session"
    }

    async fn execute(
        &self,
        args: &[String],
        env: &mut ShellEnvironment,
        _options: &ShellOptions,
    ) -> Result<i32, ShellError> {
        let mut show_pids = false;
        let mut show_running_only = false;

        // Parse arguments
        for arg in args {
            match arg.as_str() {
                "-l" => show_pids = true,
                "-r" => show_running_only = true,
                "-h" | "--help" => {
                    println!("{}", self.help());
                    return Ok(0);
                }
                _ if arg.starts_with('-') => {
                    eprintln!("jobs: {}: invalid option", arg);
                    return Ok(1);
                }
                _ => {
                    // Job ID specified - show specific job
                    if let Ok(job_id) = arg.parse::<u32>() {
                        if let Some(job) = env.get_job(job_id) {
                            print_job(job, show_pids);
                        } else {
                            eprintln!("jobs: {}: no such job", job_id);
                            return Ok(1);
                        }
                        return Ok(0);
                    } else {
                        eprintln!("jobs: {}: arguments must be job IDs", arg);
                        return Ok(1);
                    }
                }
            }
        }

        // List all jobs
        let jobs = env.get_all_jobs();
        let mut displayed = 0;

        for job in jobs {
            if show_running_only && job.status != JobStatus::Running {
                continue;
            }
            
            print_job(job, show_pids);
            displayed += 1;
        }

        if displayed == 0 && args.is_empty() {
            // No jobs to display, which is normal
        }

        Ok(0)
    }

    fn help(&self) -> String {
        "jobs [-l] [-r] [job_id ...]
Display status of jobs in the current session.

Options:
  -l    List process IDs in addition to the normal information
  -r    Display only running jobs

Arguments:
  job_id    Display information about specific job(s)

Job status indicators:
  Running   Job is currently executing
  Stopped   Job has been suspended
  Done      Job has completed

Examples:
  jobs          List all jobs
  jobs -l       List all jobs with process IDs
  jobs -r       List only running jobs
  jobs 1 2      Show details for jobs 1 and 2".to_string()
    }
}

fn print_job(job: &Job, show_pids: bool) {
    let status_char = match job.status {
        JobStatus::Running => "+",
        JobStatus::Stopped => "-",
        JobStatus::Done => " ",
    };

    let background_char = if job.background { "&" } else { "" };

    if show_pids {
        println!(
            "[{}]{} {} {} {}{}",
            job.id,
            status_char,
            job.pid,
            format!("{:?}", job.status).to_lowercase(),
            job.command,
            background_char
        );
    } else {
        println!(
            "[{}]{} {} {}{}",
            job.id,
            status_char,
            format!("{:?}", job.status).to_lowercase(),
            job.command,
            background_char
        );
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_jobs_no_jobs() {
        let mut env = ShellEnvironment::new();
        let jobs = JobsCommand;
        
        let result = jobs.execute(&[], &mut env, &ShellOptions::default()).await;
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 0);
    }

    #[tokio::test]
    async fn test_jobs_with_jobs() {
        let mut env = ShellEnvironment::new();
        
        // Add a test job
        let job = env.create_job("test command".to_string(), 123, true);
        assert_eq!(job.id, 1);

        let jobs = JobsCommand;
        let result = jobs.execute(&[], &mut env, &ShellOptions::default()).await;
        
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 0);
    }

    #[tokio::test]
    async fn test_jobs_help() {
        let mut env = ShellEnvironment::new();
        let jobs = JobsCommand;
        
        let result = jobs.execute(&["-h".to_string()], &mut env, &ShellOptions::default()).await;
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 0);
    }

    #[tokio::test]
    async fn test_jobs_invalid_option() {
        let mut env = ShellEnvironment::new();
        let jobs = JobsCommand;
        
        let result = jobs.execute(&["-x".to_string()], &mut env, &ShellOptions::default()).await;
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 1);
    }

    #[tokio::test]
    async fn test_jobs_specific_job() {
        let mut env = ShellEnvironment::new();
        
        // Add a test job
        env.create_job("test command".to_string(), 123, true);

        let jobs = JobsCommand;
        
        // Test valid job ID
        let result = jobs.execute(&["1".to_string()], &mut env, &ShellOptions::default()).await;
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 0);
        
        // Test invalid job ID
        let result = jobs.execute(&["999".to_string()], &mut env, &ShellOptions::default()).await;
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 1);
    }
}